design/reboot-interface.md

```diff
@@ -0,0 +1,190 @@
+reboot requests are managed in the same place. 
+
+If:
+
+* there is one or more annotations with the ``reboot.metal3.io`` prefix
+  present, and
+* the ``Status.PoweredOn`` field is true, and
+* the value of ``pendingRebootSince`` is empty or earlier than the
+  ``lastPoweredOn`` time
+
+then the Host controller will update ``pendingRebootSince`` to the current
+time.
+
+Whenever ``pendingRebootSince`` is later than the ``lastPoweredOn`` time, the
+Host controller will attempt to power the host off regardless of the
+``Spec.Online`` setting.
+
+Once the Host is powered off ( ``Status.PoweredOn`` is false ), if/when
+
+* the ``Spec.Online`` field is true, and
+* the ``lastPoweredOn`` time is before the ``pendingRebootSince`` time
+
+then the controller should remove the suffixless ``reboot.metal3.io``
+annotation (if present). Once no reboot annotations  are present (i.e. those of
+the form ``reboot.metal3.io/{key}`` have been removed by their originators),
+the existing logic for powering on the Host should execute and update the
+``lastPoweredOn`` timestamp accordingly.
+
+The controller automatically removes all annotations with the 
+``reboot.metal3.io`` prefix if
+* the Host is deprovisioned
+
+## Drawbacks
+
+## Future Enhancements
+
+### Defining a Formal User Interface
+
+
+### Soft Reboots
+
+The resource could accept a "soft" reboot option, in which we would attempt to
+reboot the Host without switching the power off. Since these requests may fail,
+more error reporting would likely be required. In the event that a mix of hard
+and soft reboot requests were pending, the host would be hard-rebooted and all
+of the pending requests marked complete.
+
+## Alternatives
+
+The Machine (Reboot) Actuator could perform the reboot in an imperative, rather
+than declarative, manner by manipulating the ``Online`` spec of the Host in
+sequence to ``false`` and then back to ``true`` again once the ``poweredOn``
+status was seen to be ``false``. However, in the presence of multiple actors
+this approach is prone to race conditions or controller fights.
+
+The request could simply be a timestamp field in the Host spec. The Host would
+coalesce multiple requests into one at the time they are received (rather than
+at the point of action). This would be simpler, but requires the client to
+provide a meaningful timestamp (i.e. in the recent past), forcing us to make
+awkward decisions about what to do with bogus data. In contrast, with a CRD we
+can use the creation timestamp provided by the system. This option also makes
+it more difficult to add further reboot-related API features in the future.
+
+The request could be made by adding to a list of client-chosen strings in the
+Host spec. This would be again be simpler than adding a new CRD at the cost of
+future flexibility, but requires clients to co-operate (by only adding to the
+list, and never replacing it), so one misbehaving client could cause another to
+act erratically.
```

Comment:
```suggestion
The request could be made by adding to a map of client-chosen strings in the
Host spec. This would be cleaner than using annotations, but requires clients to 
co-operate (by only adding/removing their entry in the map), so one misbehaving 
client could cause another to act erratically.
```